### PR TITLE
Fix Shadow JAR dependency publication

### LINKED ISSUE #11353: [BUG] lang-painless missing compile dependencies
**Describe the bug**
Starting `2.10.0` release the lang-painless maven package https://mvnrepository.com/artifact/org.opensearch.plugin/lang-painless/2.10.0, does not generate the compile dependencies (Maven compile dependencies are the dependencies that are required for the compilation of the project under development)

~I assume this bug was introduced  from https://github.com/opensearch-project/OpenSearch/pull/8823~, as looking at the code there are some [declared dependencies](https://github.com/opensearch-project/OpenSearch/blob/main/modules/lang-painless/build.gradle#L55-L62) but are not added in compile dependencies section of the POM file. 

**Expected behavior**
The `2.9.0` lang-painless maven package has the Compile Dependencies https://mvnrepository.com/artifact/org.opensearch.plugin/lang-painless/2.9.0

Another example of [ingest-geoip](https://mvnrepository.com/artifact/org.opensearch.plugin/ingest-geoip/2.10.0) 2.10.0 which has the compile dependencies.

### PR DESCRIPTION
**Pull request by reta**
### Description
Fix Shadow JAR dependency publication

### Related Issues
Resolves #11353

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by reta** [comment]
@prudhvigodithi I sadly cannot push to your pull request (https://github.com/opensearch-project/OpenSearch/pull/11356) but the issue was that Shadow JAR did not generate the dependencies it should be generating

**Comment by reta** [comment]
```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <modelVersion>4.0.0</modelVersion>
  <groupId>org.opensearch.plugin</groupId>
  <artifactId>lang-painless</artifactId>
  <version>3.0.0-SNAPSHOT</version>
  <name>lang-painless</name>
  <description>An easy, safe and fast scripting language for OpenSearch</description>
  <inceptionYear>2021</inceptionYear>
  <licenses>
    <license>
      <name>The Apache Software License, Version 2.0</name>
      <url>http://www.apache.org/licenses/LICENSE-2.0.txt</url>
      <distribution>repo</distribution>
    </license>
  </licenses>
  <developers>
    <developer>
      <name>OpenSearch</name>
      <url>https://github.com/opensearch-project/OpenSearch</url>
    </developer>
  </developers>
  <url>https://github.com/opensearch-project/OpenSearch</url>
  <scm>
    <url>git@github.com:opensearch-project/OpenSearch.git</url>
  </scm>
  <dependencies>
    <dependency>
      <groupId>org.antlr</groupId>
      <artifactId>antlr4-runtime</artifactId>
      <version>4.11.1</version>
      <scope>compile</scope>
      <exclusions>
        <exclusion>
          <groupId>*</groupId>
          <artifactId>*</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>org.ow2.asm</groupId>
      <artifactId>asm-util</artifactId>
      <version>9.6</version>
      <scope>compile</scope>
      <exclusions>
        <exclusion>
          <groupId>*</groupId>
          <artifactId>*</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>org.ow2.asm</groupId>
      <artifactId>asm-tree</artifactId>
      <version>9.6</version>
      <scope>compile</scope>
      <exclusions>
        <exclusion>
          <groupId>*</groupId>
          <artifactId>*</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>org.ow2.asm</groupId>
      <artifactId>asm-commons</artifactId>
      <version>9.6</version>
      <scope>compile</scope>
      <exclusions>
        <exclusion>
          <groupId>*</groupId>
          <artifactId>*</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>org.ow2.asm</groupId>
      <artifactId>asm-analysis</artifactId>
      <version>9.6</version>
      <scope>compile</scope>
      <exclusions>
        <exclusion>
          <groupId>*</groupId>
          <artifactId>*</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>org.ow2.asm</groupId>
      <artifactId>asm</artifactId>
      <version>9.6</version>
      <scope>compile</scope>
      <exclusions>
        <exclusion>
          <groupId>*</groupId>
          <artifactId>*</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>org.opensearch.plugin</groupId>
      <artifactId>opensearch-scripting-painless-spi</artifactId>
      <version>3.0.0-SNAPSHOT</version>
      <scope>compile</scope>
    </dependency>
  </dependencies>
</project>
```

**Comment by prudhvigodithi** [comment]
Thanks @reta but just curious without any change in `buildSrc/src/main/java/org/opensearch/gradle/PublishPlugin.java` it worked till 2.9.0, https://repo1.maven.org/maven2/org/opensearch/plugin/lang-painless/2.9.0/lang-painless-2.9.0.pom, the issue came up after 2.10.0 https://repo1.maven.org/maven2/org/opensearch/plugin/lang-painless/2.10.0/lang-painless-2.10.0.pom.

**Comment by reta** [comment]
> Thanks @reta but just curious without any change in `buildSrc/src/main/java/org/opensearch/gradle/PublishPlugin.java` it worked till 2.9.0, https://repo1.maven.org/maven2/org/opensearch/plugin/lang-painless/2.9.0/lang-painless-2.9.0.pom, the issue came up after 2.10.0 https://repo1.maven.org/maven2/org/opensearch/plugin/lang-painless/2.10.0/lang-painless-2.10.0.pom.

This is correct, there 2 big ones - migration to latest Nebula suite (which is what causing the problems I suspect) and updates to Shadow JAR plugin as well.

**Comment by reta** [comment]
> ❌ Gradle check result for [edfe896](https://github.com/opensearch-project/OpenSearch/commit/edfe896ba2c27e45c43542d3cbf64b66ed6ef6df): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/30521/)

https://github.com/opensearch-project/OpenSearch/issues/9191

**Comment by reta** [comment]
> ❌ Gradle check result for [edfe896](https://github.com/opensearch-project/OpenSearch/commit/edfe896ba2c27e45c43542d3cbf64b66ed6ef6df): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/30525/)

https://github.com/opensearch-project/OpenSearch/issues/10006
https://github.com/opensearch-project/OpenSearch/issues/11032

**Comment by reta** [comment]
> ❌ Gradle check result for [edfe896](https://github.com/opensearch-project/OpenSearch/commit/edfe896ba2c27e45c43542d3cbf64b66ed6ef6df): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/30529/)

https://github.com/opensearch-project/OpenSearch/issues/2775

**Comment by reta** [comment]
> ❌ Gradle check result for [edfe896](https://github.com/opensearch-project/OpenSearch/commit/edfe896ba2c27e45c43542d3cbf64b66ed6ef6df): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/30531/)

https://github.com/opensearch-project/OpenSearch/issues/10006
https://github.com/opensearch-project/OpenSearch/issues/9191

**Comment by reta** [comment]
Ended up with a single Nebula publication:

```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <!-- This module was also published with a richer model, Gradle metadata,  -->
  <!-- which should be used instead. Do not delete the following line which  -->
  <!-- is to indicate to Gradle or any Gradle module metadata file consumer  -->
  <!-- that they should prefer consuming it instead. -->
  <!-- do_not_remove: published-with-gradle-metadata -->
  <modelVersion>4.0.0</modelVersion>
  <groupId>org.opensearch.plugin</groupId>
  <artifactId>lang-painless</artifactId>
  <version>3.0.0-SNAPSHOT</version>
  <name>lang-painless</name>
  <description>OpenSearch subproject :modules:lang-painless</description>
  <dependencies>
    <dependency>
      <groupId>org.antlr</groupId>
      <artifactId>antlr4-runtime</artifactId>
      <version>4.11.1</version>
      <scope>compile</scope>
      <exclusions>
        <exclusion>
          <groupId>*</groupId>
          <artifactId>*</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>org.ow2.asm</groupId>
      <artifactId>asm-util</artifactId>
      <version>9.6</version>
      <scope>compile</scope>
      <exclusions>
        <exclusion>
          <groupId>*</groupId>
          <artifactId>*</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>org.ow2.asm</groupId>
      <artifactId>asm-tree</artifactId>
      <version>9.6</version>
      <scope>compile</scope>
      <exclusions>
        <exclusion>
          <groupId>*</groupId>
          <artifactId>*</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>org.ow2.asm</groupId>
      <artifactId>asm-commons</artifactId>
      <version>9.6</version>
      <scope>compile</scope>
      <exclusions>
        <exclusion>
          <groupId>*</groupId>
          <artifactId>*</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>org.ow2.asm</groupId>
      <artifactId>asm-analysis</artifactId>
      <version>9.6</version>
      <scope>compile</scope>
      <exclusions>
        <exclusion>
          <groupId>*</groupId>
          <artifactId>*</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>org.ow2.asm</groupId>
      <artifactId>asm</artifactId>
      <version>9.6</version>
      <scope>compile</scope>
      <exclusions>
        <exclusion>
          <groupId>*</groupId>
          <artifactId>*</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>org.opensearch.plugin</groupId>
      <artifactId>opensearch-scripting-painless-spi</artifactId>
      <version>3.0.0-SNAPSHOT</version>
      <scope>compile</scope>
    </dependency>
  </dependencies>
  <inceptionYear>2021</inceptionYear>
  <licenses>
    <license>
      <name>The Apache Software License, Version 2.0</name>
      <url>http://www.apache.org/licenses/LICENSE-2.0.txt</url>
      <distribution>repo</distribution>
    </license>
  </licenses>
  <developers>
    <developer>
      <name>OpenSearch</name>
      <url>https://github.com/opensearch-project/OpenSearch</url>
    </developer>
  </developers>
  <url>https://github.com/opensearch-project/OpenSearch</url>
  <scm>
    <url>git@github.com:opensearch-project/OpenSearch.git</url>
  </scm>
</project>

```

File: modules/lang-painless/build.gradle [hunk 1]
```diff
@@ -62,30 +61,6 @@ dependencies {
...
-  doFirst {
-    test.classpath -= project.files(project.tasks.named('shadowJar'))
-    test.classpath -= project.configurations.getByName(ShadowBasePlugin.CONFIGURATION_NAME)
-    test.classpath += project.extensions.getByType(SourceSetContainer).getByName(SourceSet.MAIN_SOURCE_SET_NAME).runtimeClasspath
-  }
-}
-
-shadowJar {
-  archiveClassifier.set('')
-  relocate 'org.objectweb', 'org.opensearch.repackage.org.objectweb'
```
Comment (reta) [inline modules/lang-painless/build.gradle:75]: We don't need that anymore since ASM is well aligned on 9.x release branch

**Review by prudhvigodithi — APPROVED** [review]

**Comment by prudhvigodithi** [comment]
Lets get this merged and backported to 2.11 so that this fix can pushed to 2.11.1 release.
@andrross @peterzhuamazon @bbarani

**Review by andrross — APPROVED** [review]

**Pull request merged**